pkts-core/src/main/java/io/pkts/packet/impl/UdpPacketImpl.java

```diff
@@ -67,7 +68,10 @@ public final void write(final OutputStream out, final Buffer payload) throws IOE
         // in this.headers after we have wrapped them. Simply wont work...
         final int size = this.headers.getReadableBytes() + (payload != null ? payload.getReadableBytes() : 0);
         this.setLength(size);
-        getParentPacket().reCalculateChecksum();
+        final IPPacket parent = getParentPacket();
+        if (parent instanceof IPv4Packet) {
+            ((IPv4Packet) parent).reCalculateChecksum();
```

Comment:
as mentioned earlier, move the checksum calculation to the IPPacket since you need to do it for both ipv4 and 6.